bittide/src/Bittide/ObliviousFifo.hs

```diff
@@ -0,0 +1,224 @@
+-- SPDX-FileCopyrightText: 2022 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+module Bittide.ObliviousFifo where
+
+import GHC.Stack
+
+import Clash.Explicit.Prelude
+import Clash.Cores.Xilinx.DcFifo hiding (DataCount)
+import qualified Clash.Cores.Extra as CE
+
+import Data.Constraint.Nat.Extra (Dict(..), leMaxLeft, leMaxRight)
+
+import Bittide.ClockControl (DataCount, targetDataCount)
+import Bittide.ElasticBuffer (EbMode(..), ebModeToReadWrite)
+import Bittide.Counter (domainDiffCounter)
+
+-- | A 'Frame' separates data frames into valid data or accumulated
+-- lost data. Furthermore, it is also used to indicate an empty buffer.
+data Frame n a =
+    Data a
+    -- ^ a single frame holding some data
+  | Empty
+    -- ^ the empty buffer indication frame
+  | Lost (Unsigned n)
+    -- ^ some number of frames whose data got lost
+  deriving (Eq, Ord, Show, Generic, NFDataX)
+
+instance (BitPack a, KnownNat n) => BitPack (Frame n a) where
```

Comment:
I just wasn't sure that deriving gets the `1 + Max n (BitSize a)` right, so I made it explict. I also wanted that `Empty` gets mapped to zero (except for the MSB), because that seems intuitive.

Reply:
Semantically there is no `Lost 0`, because you cannot loose nothing. Fun fact is that I indeed temporarly started modeling a `Positive n` type, which explicitly excludes zero from the value range. I stopped it, because it just introduces way to much overhead for just having that guarantee.

To be honest, I find the pattern solution missleading, because `Empty` signals an empty buffer, while `Lost 0` is nonsence. The pattern suggests that they somehow are equivalent.

Reply:
Yeah `Positive n` is the way to go, even if it is a bit annoying. It's non-surprising and while it does introduce syntactic overhead, it doesn't introduce any surprises.

Reply:
Yeah, that was just for the conceptual design. It now is solved differently in `Bittide.ObliviousFifo.Frame` without a custom `BitPack` instance.